File: cpp/count-square-submatrices-with-all-ones.cpp
#include "math.h"
#include "vector"

using namespace std;

class Solution {
public:
  int countSquares(vector<vector<int>>& matrix) {
    int m = matrix.size();
    int n = matrix[0].size();
    int count = 0;

    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (matrix[i][j] == 1) {
          if (i == 0 || j == 0) {
            count++;
          } else {
            matrix[i][j] = min(matrix[i - 1][j - 1], min(matrix[i - 1][j], matrix[i][j - 1])) + 1;
            count += matrix[i][j];
          }
        }
      }
    }

    return count;
  }
};

int main() {
  Solution s;

  return 0;
}

File: java/count-square-submatrices-with-all-ones.java
class Solution {
    public int countSquares(int[][] matrix) {
        int count = 0;
        int n = matrix.length;
        int m = matrix[0].length;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (matrix[i][j] == 1) {
                    if (i > 0 && j > 0) {
                        matrix[i][j] = 1 + Math.min(matrix[i - 1][j - 1], Math.min(matrix[i - 1][j], matrix[i][j - 1]));
                    }
                    count += matrix[i][j];
                }
            }
        }

        return count;
    }
}